src/main/java/software/amazon/nio/spi/s3/config/S3NioSpiConfiguration.java

```diff
@@ -27,86 +38,271 @@ public class S3NioSpiConfiguration {
+            .forEach(key -> put(key, overrides.getProperty(key)));
+    }
+
+    /**
+     * Fluently sets the value of maximum fragment number
+     *
+     * @param maxFragmentNumber the maximum fragment number
+     *
+     * @return this instance
+     */
+    public S3NioSpiConfiguration withMaxFragmentNumber(int maxFragmentNumber) {
+        if (maxFragmentNumber < 1) {
+            throw new IllegalArgumentException("maxFragmentNumber must be positive");
+        }
+        put(S3_SPI_READ_MAX_FRAGMENT_NUMBER_PROPERTY, String.valueOf(maxFragmentNumber));
+        return this;
+    }
+
+    /**
+     * Fluently sets the value of maximum fragment size
+     *
+     * @param maxFragmentSize the maximum fragment size
+     *
+     * @return this instance
+     */
+    public S3NioSpiConfiguration withMaxFragmentSize(int maxFragmentSize) {
+        if (maxFragmentSize < 1) {
+            throw new IllegalArgumentException("maxFragmentSize must be positive");
+        }
+        put(S3_SPI_READ_MAX_FRAGMENT_SIZE_PROPERTY, String.valueOf(maxFragmentSize));
+        return this;
+    }
+
+    /**
+     * Fluently sets the value of the endpoint
+     *
+     * @param endpoint the endpoint
+     *
+     * @return this instance
+     */
+    public S3NioSpiConfiguration withEndpoint(String endpoint) {
+
+        if (endpoint == null) {
+            endpoint = "";
+        }
+        endpoint = endpoint.trim();
+
+        if ((endpoint.length() > 0) && !ENDPOINT_REGEXP.matcher(endpoint).find()) {
+            throw new IllegalArgumentException(
+                String.format("endpoint '%s' does not match format host:port", endpoint)
+            );
+        }
+
+        put(S3_SPI_ENDPOINT_PROPERTY, endpoint); return this;
+    }
+
+    /**
+     * Fluently sets the value of the endpoint's protocol
+     *
+     * @param protocol the endpoint's protcol
+     *
+     * @return this instance
+     */
+    public S3NioSpiConfiguration withEndpointProtocol(String protocol) {
+        if (protocol != null) {
+            protocol = protocol.trim();
+        }
+        if (!"http".equals(protocol) && !"https".equals(protocol)) {
+            throw new IllegalArgumentException("endpoint prococol must be one of ('http', 'https')");
+        }
+        put(S3_SPI_ENDPOINT_PROTOCOL_PROPERTY, protocol); return this;
+    }
+
+    /**
+     * Fluently sets the value of the region
+     *
+     * @param region the region; if null or blank the property is removed
+     *
+     * @return this instance
+     */
+    public S3NioSpiConfiguration withRegion(String region) {
+        if ((region == null) || StringUtils.isBlank(region)) {
+            remove(AWS_REGION_PROPERTY);
+        } else {
+            put(AWS_REGION_PROPERTY, region.trim());
+        }
+
+        return this;
+    }
+
+    /**
+     * Fluently sets the value of accessKey and secretAccessKey
+     *
+     * @param accessKey the accesskey; if null, credentials are removed
```

Comment:
Could this method just take an instance of https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/auth/credentials/AwsCredentials.html

Reply:
Yep, that could be an option; I actually consider it, but look at the required code from a user perspective:

option1: config.withCredentials("key", "secret");
option2: config.withCredentials(AwsBasicCredentials.create("key", "secret")); (plus the import)

I would prefer option1 :) - although we can have an additional withCredentials(cred) in the case cred is already a AwsCredentials().

Reply:
Using access key and secret access key are but one way to create credentials in AWS. There are also situations in which you would need to provide additional information such as when using temporary credentials. The AWS docs list the full range of environment variables which can be used. I would therefore remove this version of `withCredentials` and use only `withCredentials(AwsCredentials)` and store that object in the map if needed.


If desired, you could retain this version of the method as a convenience method but use it to build an `AwsBasicCredentials` object and then use that in `withCredentials(credentials). An inversion of what you have here.

Reply:
done